facechain/train_text_to_image_paiya.py

```diff
@@ -0,0 +1,1297 @@
+    # 然后对真人图片取mean，获取真人图片的平均特征
+    pivot_feature   = np.mean(embedding_array, axis=0)
+    pivot_feature   = np.reshape(pivot_feature, [512, 1])
+
+    # 计算一个文件夹中，和中位值最接近的图片排序
+    scores = [np.dot(emb, pivot_feature)[0][0] for emb in embedding_list]
+    return scores
+
+def prepare_dataset_paiya(instance_images: list, output_dataset_dir, work_dir):
+    from .utils.face_process_utils import call_face_crop
+    # 创建输出文件夹
+    images_save_path    = os.path.join(output_dataset_dir, "train")
+    json_save_path      = os.path.join(output_dataset_dir, "metadata.jsonl")
+    best_roop_image_path = os.path.join(work_dir, "best_outputs")
+    os.makedirs(images_save_path, exist_ok=True)
+    os.makedirs(best_roop_image_path, exist_ok=True)
+
+    # 人脸评分
+    face_quality_func       = pipeline(Tasks.face_quality_assessment, 'damo/cv_manual_face-quality-assessment_fqa')
+    # embedding
+    face_recognition        = pipeline(Tasks.face_recognition, model='damo/cv_ir101_facerecognition_cfglint')
+    # 人脸检测
+    retinaface_detection    = pipeline(Tasks.face_detection, 'damo/cv_resnet50_face-detection_retinaface')
+    # 显著性检测
+    salient_detect          = pipeline(Tasks.semantic_segmentation, model='damo/cv_u2net_salient-detection')
+    
+    # 获得jpg列表
+    jpgs            = instance_images
+    # ---------------------------人脸得分计算-------------------------- #
+    face_id_scores  = []
+    quality_scores  = []
+    face_angles     = []
+    selected_paths  = []
+    for index, jpg in enumerate(tqdm(jpgs)):
+        try:
+            if not jpg.lower().endswith(('.bmp', '.dib', '.png', '.jpg', '.jpeg', '.pbm', '.pgm', '.ppm', '.tif', '.tiff')):
+                continue
+            image       = Image.open(jpg)
+            h, w, c     = np.shape(image)
+
+            retinaface_box, retinaface_keypoint, _ = call_face_crop(retinaface_detection, image, 3, prefix="tmp")
+            retinaface_keypoint = np.reshape(retinaface_keypoint, [5, 2])
+            # 计算人脸偏移角度
+            x = retinaface_keypoint[0,0] - retinaface_keypoint[1,0]
+            y = retinaface_keypoint[0,1] - retinaface_keypoint[1,1]
+            angle = 0 if x==0 else abs(math.atan(y/x)*180/math.pi)
+            angle = (90 - angle)/ 90 
+
+            # 人脸宽度判断
+            face_width  = (retinaface_box[2] - retinaface_box[0]) / (3 - 1)
+            face_height = (retinaface_box[3] - retinaface_box[1]) / (3 - 1)
+            if face_width / w < 1/8 or face_height / h < 1/8:
+                continue
+
+            sub_image = image.crop(retinaface_box)
+
+            embedding   = np.array(face_recognition(sub_image)[OutputKeys.IMG_EMBEDDING])
+            score       = face_quality_func(sub_image)[OutputKeys.SCORES]
+            score       = 0 if score is None else score[0]
+
+            face_id_scores.append(embedding)
+            quality_scores.append(score)
+            face_angles.append(angle)
+
+            selected_paths.append(jpg)
+        except:
+            pass
```

Comment:
face pick module should be extracted to be used by developer more conviniently

Reply:
i will be merged in preprocessing.py